Moved `IteHandler` into the preprocessing pipeline

Tomaqa opened the pull request:
Fixes `TermNames` to also consider the mapping of internal terms to original user terms.
So far, I have only identified `IteHandler` to change the `PTRef` of the inserted formula in `insertFormula` and at the same time store it into the assertions, so that the changed `PTRef` is retrieved via getting the assertions or via `PartitionaManager`.
Without this, the `UnsatCoreBuilder` would consider some terms to be unnamed because it would not recognize them.

Resolves #857

Tomaqa committed:
Regression tests on unsat cores with terms that may internally be changed

blishko reviewed (commented):
Seems good, but I am not sure I fully understand the underlying problem.
There should be some very minimal example to reproduce the problem, no?
Can we find it and use it as the regression test, instead of these large files?
It would also help me to understand the problem properly.
`ITEHandler` has always been a bit of a hack.
Wouldn't it help if we resolved the TODO associated with its usage, and incorporate the `ite` processing to the preprocessing pipeline?

Tomaqa commented:
> There should be some very minimal example to reproduce the problem, no?

I identified some very small examples and added them. I believe that the problem appears when the top-level assertion is an ITE. I will also try to make some similar unit tests.

Tomaqa committed:
Unit tests on unsat cores with top-level ITE terms that are internally changed

Tomaqa commented:
I also included unit tests. As a consequence, I also added to possibility to retrieve the names from `NamedUnsatCore` and added several tests on this in the unit tests.

Even with the fix, we correctly handle the unsat cores such that the names are kept, but not the terms themselves. See e.g. `test_UnsatCore.cc:395` (in the `Bool_Trivial_ITE` test).
Hence, for example, in the case of full cores where the names are ignored, the user cannot track which of the terms he/she asserted are kept/discarded from the core. That does not seem right, but I can fix it by storing the original terms instead of the internals into the core.

Tomaqa commented:
My current feeling is that the transformation of internal terms to the original should already happen inside of `api/PartitionManager.h:51` :
```
inline std::vector<PTRef> getPartitions() const { return partitionInfo.getTopLevelFormulas(); }
```

I am not sure if this is always the desired behavior, or if it should be optional - a parameter in `PartitionManager` would allow to do the mapping.
But since `getPartitions()` refers to top-level assertions, I think it should refer to the original `PTRef` provided to the `insertFormula`.

Tomaqa commented:
I checked that the current fix resolves all the issues reported in #857, including minimal cores, except the cases reported in #859 where the issue is probably not related to the one discussed here.

Tomaqa committed:
Moved preprocessing of ITE terms into the preprocessing pipeline

Tomaqa commented:
> `ITEHandler` has always been a bit of a hack. Wouldn't it help if we resolved the TODO associated with its usage, and incorporate the `ite` processing to the preprocessing pipeline?

I moved the use of the `IteHandler` into preprocessing, and it seems to resolve the issues.
Now, the partition manager knows the mapping and should always return the original `PTRef`s - which comprises both unsat cores and interpolants, which both could have returned internal `PTRef`, thus unknown to the user.

blishko commented on src/api/MainSolver.cc:
```diff
@@ -172,7 +173,13 @@ sstat MainSolver::simplifyFormulas() {
+                frameFormulas.push(fla);
+            }
+
+            PTRef frameFormula = logic.mkAnd(std::move(frameFormulas));
```
Why not apply IteHandler once on the conjunction?
Wouldn't that be better?

Tomaqa replied:
I have not realized that it works recursively. Yes, it's better!

blishko commented on test/regression/unsatcores/QF_LIA/calypto-2.smt2:
Do we need this file in tests?
One, the core issue should be caught by the smaller tests, no?
Two, this one seems brittle, if the minimal unsat core is just `smtcomp43` it seems weird to assert that OpenSMT will always find the unsat core consisting of 
```

(
smtcomp14
smtcomp21
smtcomp43
)
```

blishko reviewed (commented):
I have to refresh my memory why IteHandler needs a number.
I am not sure about the `simplifiedFormulasCount`.

What I don't quite like is that now, the same original formula can be preprocessed with IteHandler multiple times.
I am afraid this could lead to some performance regression in incremental solving.

Tomaqa committed:
Explicitly including used STL headers in MainSolver.h

Tomaqa committed:
Refactored preprocessing of ITE terms into separate function

Tomaqa committed:
Do not postpone preprocessing of ITEs when UF is combined with LA

Tomaqa commented:
> > What I don't quite like is that now, the same original formula can be preprocessed with IteHandler multiple times.
> 
> I will need to handle this more thoroughly. I think it would be best to first implement an array of counters ...

Counters implemented in #889.

Now `IteHandler` is not being called multiple times: if `trackPartitions()`, then it is guaranteed thanks to the counters, not preprocessing already preprocessed formulas. In the other case, all frame formulas are always put into conjunction, potentially preprocessing some formulas again. We can use caching, but I realized that since the conjunctions are always different, there may never be a cache hit. We may instead separately remember the previous preprocessed formula and use it, and apply `IteHandler` only on the new part of the formula.

Tomaqa committed:
Making sure IteHandler is not called repeatedly on the already seen formulas

Tomaqa committed:
Added cache for the preprocessing of ITE terms
